### PR TITLE
Tree-LSTM example

### PR DESCRIPTION
**Pull request by zhisbug**
#1288 
@neubig at your convenience please take a look at the dropout implementation, which seems to be extremely slow.

Reported by @zhiyong1997 : Without dropout, one epoch takes ~11 seconds on GPU, but when one dropout is added (only to the layer before logits calculation), the time becomes ~18 seconds. By contrast, when three different dropouts are added in TF fold at the same time, the time only turns from ~13s to ~15s, which is ~30s for dynet.

### CODE REVIEW

**Comment by neubig** [comment]
This is because the dropout node is not being autobatched because it doesn't have an autobatching profile. I'll work on adding one today.

**Comment by zhisbug** [comment]
Ready to be merged. @xunzhang will manually squash&merge in order to keep the authorship of commits.
Thanks @zhiyong1997 for the great contribution.

**Comment by xunzhang** [comment]
Squashed and merged into master https://github.com/clab/dynet/commit/2bbc0b965febd5422e0dbbb68dc96da1e3f8dfdd and https://github.com/clab/dynet/commit/0eddde2f9a47e3a9fd98287d4590b5b1391a479b

**Comment by xunzhang** [comment]
@zhiyong1997 please also fix the Codacy Quality check in the future contribution.
